File: src/ch14/markov_gen.pas
(*
Python Cookbook, 3rd Ed.

Chapter 14, Application Integration: Combination
Markov Generator
*)
program markov_gen;
uses
    sysUtils,
    Dos;

type
    (* Command-line options. *)
    Options = record
        file_name: String;
        sample_count: Integer;
        randomize: Cardinal;
    end;

    (* The Markov Chain and final outcome state. *)
    Outcome = (InProcess, Success, Fail);
    Chain = record
        target: SmallInt;
        len:  SmallInt;
        sequence:  array of Shortint;
        outcome: Outcome;
    end;

    WriteHeader = procedure(var out: Text; opts: Options; header: Boolean);

    WriteSample = procedure(var out: Text; ch: Chain);

function ProbeValue(): ShortInt;
begin
    Probevalue := Random(6) + Random(6) + 2;
end;

procedure Append(var probe: ShortInt; var ch: Chain);
begin
    SetLength(ch.sequence, ch.len+1);
    ch.sequence[ch.len] := probe;
    ch.len := ch.len + 1;
end;

function Start(probe: ShortInt; var ch: Chain) : Boolean;
begin
    Append(probe, ch);
    case probe of
        7, 11: begin
            ch.outcome := Success;
            ch.target := -1;
            Start := False;
            end;
        2, 3, 12: begin
            ch.outcome := Fail;
            ch.target := -1;
            Start := False;
            end;
    else begin
        ch.outcome := InProcess;
        ch.target := probe;
        Start := True;
        end;
    end;
end;

function GrowUntil(probe: ShortInt; var ch: Chain) : Boolean;
begin
    Append(probe, ch);
    if probe = 7 then begin
        ch.outcome := Fail;
        GrowUntil := False;
        end
    else if probe = ch.target then begin
        ch.outcome := Success;
        GrowUntil := False;
        end
    else
        GrowUntil := True;
end;

function MakeChain: Chain;
var
    ch : Chain;
    continue : Boolean;
    pv : ShortInt;
begin
    SetLength(ch.sequence, 0);
    ch.len := 0;
    pv := ProbeValue();
    continue := Start(pv, ch);
    while continue do begin
        pv := ProbeValue();
        continue := GrowUntil(pv, ch);
        end;
    MakeChain := ch;
end;

procedure ParseArgs(var opts: Options);
var
    param: LongInt;
    option: String;
begin
    opts.sample_count := 100;
    opts.file_name := '';  (* Will write to StdOut. *)
    if GetEnv('RANDOMSEED') <> '' then
        opts.randomize := StrToInt(GetEnv('RANDOMSEED'))
    else
        opts.randomize := 0;
    param := 1;
    while param <= ParamCount do begin
        if ParamStr(param)[1] = '-' then begin
            option := ParamStr(param);
            Delete(option, 1, 1);
            LowerCase(option);
            case option of
                's', '-samples': begin
                    param := param + 1;
                    opts.sample_count := StrToInt(ParamStr(param));
                    end;
                'r', '-randomize': begin
                    param := param + 1;
                    opts.randomize := StrToInt(ParamStr(param));
                    end;
                'o', '-output': begin
                    param := param + 1;
                    opts.file_name := ParamStr(param);
                    end;
                'h', '-help': begin
                    WriteLn('usage: ./markov --samples n --randomize s [output]');
                    WriteLn();
                    WriteLn('Markov Chain Generator');
                    WriteLn();
                    WriteLn('-s --samples n     The number of samples to generate, default is 100.');
                    WriteLn('-r --randomize s   Seed for Random generator.');
                    WriteLn('-o --output name   Ouput file name, default writes to stdout.');
                    WriteLn;
                    WriteLn('Output is a TOML-format file with two tables: Configuration and Samples.');
                    WriteLn('The Samples table is an array of {outcome="...", chain=[...]} mappings.');
                    Halt(1);
                    end;
            else begin
                WriteLn('Invalid option ', ParamStr(param));
                Halt(2);
                end
            end
        end
        else begin
            WriteLn('Invalid positional parameter ', ParamStr(param));
            Halt(2);
        end;
        param := param + 1;
    end;
end;

procedure TOMLWriteHeader(var out: Text; opts: Options; columns: Boolean);
begin
    WriteLn(out, '[Configuration]');
    WriteLn(out, '  file = "', opts.file_name, '"');
    WriteLn(out, '  samples = ', opts.sample_count);
    Writeln(out, '  randomize = ', opts.randomize);
end;

procedure TOMLWriteSample(var out: Text; ch: Chain);
var
    I: Integer;
begin
    WriteLn(out, '[[Samples]]');
    WriteLn(out, '  outcome="', ch.outcome, '"');
    WriteLn(out, '  length=', ch.len);
    Write(out, '  chain=[', ch.sequence[0]);
    For I := 1 to ch.len-1 do
        Write(out, ', ', ch.sequence[I]);
    WriteLn(out, ']');
end;

procedure CSVWriteHeader(var out: Text; opts: Options; columns: Boolean);
begin
    WriteLn(out, '# file = "', opts.file_name, '"');
    WriteLn(out, '# samples = ', opts.sample_count);
    Writeln(out, '# randomize = ', opts.randomize);
    if columns then begin
        Writeln(out, '# -----');
        Writeln(out, 'outcome,length,chain');
    end;
end;

procedure CSVWriteSample(var out: Text; ch: Chain);
var
    I: Integer;
begin
    Write(out, '"', ch.outcome, '"');
    Write(out, ',', ch.len);
    Write(out, ',"', ch.sequence[0]);
    For I := 1 to ch.len-1 do
        Write(out, ';', ch.sequence[I]);
    WriteLn(out, '"');
end;

var
    opts: Options;
    ch: Chain;
    out: Text;
    count: Integer;

    wtrHeader: WriteHeader;
    wtrSample: WriteSample;

begin
    ParseArgs(Opts);

    (* Based on file_filename; default to CSV when writing to StdOut. *)
    if RightStr(opts.file_name, 5) = '.toml' then begin
        wtrHeader := @TOMLWriteHeader;
        wtrSample := @TOMLWriteSample;
    end
    else begin
        wtrHeader := @CSVWriteHeader;
        wtrSample := @CSVWriteSample;
    end;

    if Opts.randomize <> 0 then
        RandSeed := opts.randomize
    else
        Randomize;

    Assign(out, opts.file_name);
    Rewrite(out);

    wtrHeader(out, opts, True);

    for count := 1 to opts.sample_count do begin
        ch := MakeChain();
        wtrSample(out, ch);
    end;

    Close(out);

    if Opts.file_name <> '' then begin
        wtrHeader(StdOut, opts, False);
    end;
end.

File: src/ch14/markov_gen.py
"""
Python Cookbook, 3rd Ed.

Chapter 14, Application Integration: Combination
Markov Generator
"""
import argparse
from collections.abc import Callable
from contextlib import redirect_stdout
import csv
from functools import partial
import os
from pathlib import Path
import random
import sys
from textwrap import dedent
from typing import TypeAlias, Any, TextIO

Chain: TypeAlias = list[int]
State: TypeAlias = Callable[[Chain, int], tuple[Chain, Any]]


class Succeed(Exception):
    pass


class Fail(Exception):
    pass


def start(chain: Chain, roll: int) -> tuple[Chain, State]:
    if roll in {7, 11}:
        return chain + [roll], succeed
    elif roll in {2, 3, 12}:
        return chain + [roll], fail
    else:
        return chain + [roll], partial(grow_until, roll)


def grow_until(point: int, chain: Chain, roll: int) -> tuple[Chain, State]:
    if roll in {7}:
        return chain + [roll], fail
    elif roll == point:
        return chain + [roll], succeed
    else:
        return chain + [roll], partial(grow_until, point)


def succeed(chain: Chain, roll: int) -> tuple[Chain, State]:
    raise Succeed(chain)


def fail(chain: Chain, roll: int) -> tuple[Chain, State]:
    raise Fail(chain)


class Dice:
    def __init__(self, seed: int | None = None) -> None:
        self._rng = random.Random(seed)
        self.roll()

    def roll(self) -> tuple[int, ...]:
        self.dice = (self._rng.randint(1, 6), self._rng.randint(1, 6))
        return self.dice


def make_chain(dice: Dice) -> tuple[str, Chain]:
    state: State = start
    chain: Chain = []
    try:
        while True:
            roll = sum(dice.roll())
            chain, state = state(chain, roll)
    except Succeed:
        return ("Success", chain)
    except Fail:
        return ("Fail", chain)


class Writer:
    def __init__(self, target: TextIO | None = None) -> None:
        self.target = target

    def header(self, opts: argparse.Namespace, columns: bool = True) -> None:
        ...

    def sample(self, outcome: str, chain: list[int]) -> None:
        ...


class CSVWriter(Writer):
    def __init__(self, target: TextIO | None = None) -> None:
        super().__init__(target)

    def header(self, opts: argparse.Namespace, columns: bool = True) -> None:
        print(f'# file = "{opts.output}"', file=self.target)
        print(f"# samples = {opts.samples}", file=self.target)
        print(f"# randomize = {opts.randomize}", file=self.target)
        if columns and self.target:
            print("# -----", file=self.target)
            self.writer = csv.writer(self.target)
            self.writer.writerow(["outcome", "length", "chain"])

    def sample(self, outcome: str, chain: list[int]) -> None:
        self.writer.writerow([outcome, len(chain), ";".join(map(str, chain))])


class TOMLWriter(Writer):
    def __init__(self, target: TextIO | None = None) -> None:
        super().__init__(target or sys.stdout)

    def header(self, opts: argparse.Namespace, columns: bool = True) -> None:
        with redirect_stdout(self.target):
            print("[Configuration]")
            print(f'  file = "{opts.output}"')
            print(f"  samples = {opts.samples}")
            print(f"  randomize = {opts.randomize}")

    def sample(self, outcome: str, chain: list[int]) -> None:
        with redirect_stdout(self.target):
            print("[[Samples]]")
            print(f'  outcome = "{outcome}"')
            print(f"  length = {len(chain)}")
            print(f"  chain = {chain}")


def write_samples(target_file: TextIO, opts: argparse.Namespace) -> None:
    if opts.randomize:
        probe_sequence = Dice(opts.randomize)
    else:
        probe_sequence = Dice()

    writer = CSVWriter(target_file)
    writer.header(opts, columns=True)
    for i in range(opts.samples):
        outcome, chain = make_chain(probe_sequence)
        writer.sample(outcome, chain)


def get_options(argv: list[str]) -> argparse.Namespace:
    try:
        default_seed = int(os.environ.get("RANDOMSEED", 0))
    except ValueError:
        default_seed = 0
    parser = argparse.ArgumentParser(description="Markov Chain Generator")
    parser.add_argument("-s", "--samples", type=int, default=100)
    parser.add_argument("-r", "--randomize", type=int, default=default_seed)
    parser.add_argument("-o", "--output", type=Path, default=None)
    options = parser.parse_args(argv)
    return options


def main(argv: list[str] = sys.argv[1:]) -> None:
    options = get_options(argv)

    if options.output:
        with options.output.open("w") as target_file:
            write_samples(target_file, options)
        # Summary
        writer = CSVWriter()
        writer.header(options, columns=False)
    else:
        write_samples(sys.stdout, options)


if __name__ == "__main__":
    main()
